test/e2e/set/scyllacluster/scyllacluster_shardawareness.go

```diff
@@ -69,57 +66,131 @@ var _ = g.Describe("ScyllaCluster", func() {
 		o.Expect(err).NotTo(o.HaveOccurred())
 		o.Expect(hosts).To(o.HaveLen(1))
 
-		connections := make(map[uint16]string)
-		var connectionsMut sync.Mutex
-
-		clusterConfig := gocql.NewCluster(hosts...)
-		clusterConfig.Dialer = DialerFunc(func(ctx context.Context, network, addr string) (net.Conn, error) {
-			sourcePort := gocql.ScyllaGetSourcePort(ctx)
-			localAddr, err := net.ResolveTCPAddr(network, fmt.Sprintf(":%d", sourcePort))
-			if err != nil {
-				return nil, err
-			}
-
-			framework.Infof("Connecting to %s using %d source port", addr, sourcePort)
-			connectionsMut.Lock()
-			connections[sourcePort] = addr
-			connectionsMut.Unlock()
+		clientPod := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "client",
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:  "client",
+						Image: configassests.Project.OperatorTests.NodeSetupImage,
+						Command: []string{
+							"/bin/sh",
+							"-c",
+							"sleep 3600",
+						},
+					},
+				},
+				TerminationGracePeriodSeconds: pointer.Int64(1),
+				RestartPolicy:                 corev1.RestartPolicyNever,
+			},
+		}
+		clientPod, err = f.KubeClient().CoreV1().Pods(f.Namespace()).Create(ctx, clientPod, metav1.CreateOptions{})
 
-			d := &net.Dialer{LocalAddr: localAddr}
-			return d.DialContext(ctx, network, addr)
-		})
+		waitCtx2, waitCtx2Cancel := utils.ContextForPodStartup(ctx)
+		defer waitCtx2Cancel()
+		clientPod, err = controllerhelpers.WaitForPodState(waitCtx2, f.KubeClient().CoreV1().Pods(clientPod.Namespace), clientPod.Name, controllerhelpers.WaitForStateOptions{}, utils.PodIsRunning)
 
-		framework.By("Waiting for the driver to establish connection to shards")
-		session, err := gocqlx.WrapSession(clusterConfig.CreateSession())
-		o.Expect(err).NotTo(o.HaveOccurred())
-		defer session.Close()
+		const (
+			// Ref: https://github.com/apache/cassandra/blob/cassandra-5.0/doc/native_protocol_v4.spec
+			optionsFrame   = `\x04\x00\x00\x00\x05\x00\x00\x00\x00`
+			scyllaShardKey = "SCYLLA_SHARD"
 
-		err = wait.PollUntilContextTimeout(ctx, 100*time.Millisecond, 5*time.Second, true, func(context.Context) (done bool, err error) {
-			return len(connections) == nrShards, nil
-		})
-		o.Expect(err).NotTo(o.HaveOccurred())
+			connectionAttempts = 10
+		)
 
-		shardAwareAttempts := 0
-		for sourcePort, addr := range connections {
-			// Control connection is also put in pool, and it always uses the default port.
-			if sourcePort == 0 {
-				o.Expect(addr).To(o.HaveSuffix(fmt.Sprintf("%d", nonShardAwarePort)))
-				continue
+		for shard := 0; shard < nrShards; shard++ {
```

Comment:
nit (or rather a tiny hint): since 1.22 you can just do `for shard := range nrShards { ... }` for range over integers

Reply:
done